### PR TITLE
[SPARK-17488][CORE] TakeAndOrder will OOM when the data is very large

### PR DESCRIPTION
**Pull request by cenyuhai**
## What changes were proposed in this pull request?

In function Utils.takeOrdered, it will sort all data in memory, when the data is very large, It will OOM. This pr is to add external sorter for function takeOrdered.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/rdd/RDD.scala [hunk 1]
```diff
@@ -493,8 +494,7 @@ abstract class RDD[T: ClassTag](
    *
    * @param weights weights for splits, will be normalized if they don't sum to 1
    * @param seed random seed
-   *
-   * @return split RDDs in an array
+    * @return split RDDs in an array
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala]: Can you revert these unnecessary changes?

Comment (cenyuhai) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala]: I am so sorry for this. I will revert it later.

File: core/src/main/scala/org/apache/spark/util/collection/Utils.scala [hunk 2]
```diff
@@ -30,10 +31,15 @@ private[spark] object Utils {
...
    * and maintains the ordering.
    */
-  def takeOrdered[T](input: Iterator[T], num: Int)(implicit ord: Ordering[T]): Iterator[T] = {
-    val ordering = new GuavaOrdering[T] {
-      override def compare(l: T, r: T): Int = ord.compare(l, r)
-    }
-    ordering.leastOf(input.asJava, num).iterator.asScala
+  def takeOrdered[T](input: Iterator[T], num: Int,
+      ser: Serializer = SparkEnv.get.serializer)(implicit ord: Ordering[T]): Iterator[T] = {
+    val context = TaskContext.get()
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/util/collection/Utils.scala]: `context` could be null if task context is not set. Better to handle it.

**Comment by viirya** [comment]
cc @cloud-fan @hvanhovell

File: core/src/main/scala/org/apache/spark/util/collection/Utils.scala [hunk 3]
```diff
@@ -30,10 +34,22 @@ private[spark] object Utils {
...
+    if (context == null) {
+      val ordering = new GuavaOrdering[T] {
+        override def compare(l: T, r: T): Int = ord.compare(l, r)
+      }
+      ordering.leastOf(input.asJava, num).iterator.asScala
+    } else {
+      val sorter =
+        new ExternalSorter[T, Any, Any](context, None, None, Some(ord), ser)
+      sorter.insertAll(input.map(x => (x, null)))
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/util/collection/Utils.scala]: Can `GuavaOrdering` sort the data more than the memory capacity? I quickly check its source code, looks it can't, but not very sure.

Comment (viirya) [inline core/src/main/scala/org/apache/spark/util/collection/Utils.scala]: I was thinking the value of this PR is to sort huge data and return small top k. If `GuavaOrdering` can already do this, I would suspect the value of this change. @cenyuhai What do you think?

Comment (srowen) [inline core/src/main/scala/org/apache/spark/util/collection/Utils.scala]: Well, it only resorts to sorting when _k_ is more than `Int.MaxValue / 2` because the normal algorithm can't run. This is an extreme case, where "take" isn't very sensible to begin with.

Generally, you would never want to sort the elements to get the top k. The Guava class is doing what the `BoundedPriorityQueue` internal class does, and keeps a priority queue of the top k and only sorts at the end.

I think the current approach is much better than sorting everything, and because you have to return an array of k elements in several cases, it's not really solving the OOM problem.

File: core/src/main/scala/org/apache/spark/rdd/RDD.scala [hunk 4]
```diff
@@ -1384,14 +1385,15 @@ abstract class RDD[T: ClassTag](
    * @param ord the implicit ordering for T
    * @return an array of top elements
    */
-  def takeOrdered(num: Int)(implicit ord: Ordering[T]): Array[T] = withScope {
+  def takeOrdered(num: Int, ser: Serializer = SparkEnv.get.serializer)
```
Comment (cenyuhai) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala]: Yes, you are right.

File: core/src/main/scala/org/apache/spark/util/collection/Utils.scala (hunk 3 above)
Comment (cenyuhai) [inline core/src/main/scala/org/apache/spark/util/collection/Utils.scala]: 1.In my case,  user execute a sql "select * from table sort by time limit 10000000", the k is very large, it's an extreme case. I need not change RDD.takeOrdered. I will limit the changes in limit.scala.
2. GuavaOrdering will sort all data in memory and then take top k. If there is enough memory, ExternalSorter will not spill.

**Comment by srowen** [comment]
This is introducing a lot of complexity and I'm still not sure why we'd want to _sort_ here. You do not in general want to sort something to get the result here.

**Comment by cenyuhai** [comment]
In terms of data security, we can't get all data. Every sql should limit 10 million records. But sometimes it will OOM... My demand is to avoid OOM. @srowen Do you have any idea?

**Comment by srowen** [comment]
This takes a heavy sort-based path well before it would cause an OOM it seems though. This is a simple Utils method and I'm not sure it should be causing spills to disk. If the problem is that you're running out of memory taking 10M records to the driver, don't you just generally need more driver memory? you're going to be in memory eventually on the driver when the result is produced.

**Comment by cenyuhai** [comment]
Driver is ok, but executor is running out of memory, this method is called by executor. Our maximum limit of memory is 15G.

**Comment by cloud-fan** [comment]
Limiting 10 million records is really a corner case. A workaround is to save your sorted query result into a new table and read the new table with your large limit. I don't think this corner case worth introducing the complexity to open source code base.

**Comment by cenyuhai** [comment]
Ok, I close this PR
